crates/pacdef_core/src/backend/actual/rustup.rs

```diff
@@ -0,0 +1,231 @@
+
+impl Backend for Rustup {
+    impl_backend_constants!();
+
+    fn get_all_installed_packages(&self) -> Result<HashSet<Package>> {
+        let mut toolchains_vec = self
+            .run_toolchain_command(get_info_switches(Repotype::Toolchain))
+            .context("Getting installed toolchains")?;
+
+        let mut toolchains: HashSet<Package> = toolchains_vec
+            .iter()
+            .map(|name| ["toolchain", name].join("/").into())
+            .collect();
+
+        let packages: HashSet<Package> = self
+            .run_component_command(get_info_switches(Repotype::Component), &mut toolchains_vec)
+            .context("Getting installed components")?
+            .iter()
+            .map(|name| ["component", name].join("/").into())
+            .collect();
+
+        toolchains.extend(packages);
+        Ok(toolchains)
+    }
+
+    fn get_explicitly_installed_packages(&self) -> Result<HashSet<Package>> {
+        self.get_all_installed_packages()
+            .context("Getting all installed packages")
+    }
+
+    fn make_dependency(&self, _: &[Package]) -> Result<ExitStatus> {
+        anyhow::bail!("Not supported by {}", self.get_binary())
+    }
+
+    fn install_packages(&self, packages: &[Package], _: bool) -> Result<ExitStatus> {
+        for p in packages {
+            let repo = if p.repo.is_some() {
+                p.repo.as_ref().expect("This should never be reached")
+            } else {
+                anyhow::bail!("Not specified whether it is a toolchain or a component")
+            };
+
+            let mut cmd = Command::new(self.get_binary());
+            match repo.as_str() {
+                "toolchain" => {
+                    cmd.args(get_install_switches(Repotype::Toolchain));
+                    cmd.arg(&p.name);
+                }
+                "component" => {
+                    cmd.args(get_install_switches(Repotype::Component));
+                    let mut iter = p.name.split('/');
+                    let toolchain = iter.next().expect("Toolchain not specified!");
+                    let component = iter.next().expect("Component not specified!");
+                    cmd.arg(toolchain).arg(component);
+                }
+                _ => anyhow::bail!("No such type is managed by rustup!"),
+            }
+
+            let result = cmd.status().context("Installing toolchain {p}");
+            if !result.as_ref().is_ok_and(|exit| exit.success()) {
+                return result;
+            }
+        }
+        Ok(ExitStatus::from_raw(0))
+    }
+
+    fn remove_packages(&self, packages: &[Package], _: bool) -> Result<ExitStatus> {
+        let mut toolchains_rem = Vec::new();
+
+        for p in packages {
+            let repo = if p.repo.is_some() {
+                p.repo.as_ref().expect("This will never be printed!")
+            } else {
+                anyhow::bail!("Not specified whether it is a toolchain or a component")
+            };
```

Comment:
This is a `match` statement in disguise.

Reply:
Ah my bad. Wasn't thinking :sweat_smile: